plugin/manager_linux.go

```diff
@@ -146,26 +146,35 @@ func (pm *Manager) restore(p *v2.Plugin, c *controller) error {
 	return nil
 }
 
+const shutdownTimeout = 10 * time.Second
+
 func shutdownPlugin(p *v2.Plugin, ec chan bool, executor Executor) {
 	pluginID := p.GetID()
 
 	err := executor.Signal(pluginID, int(unix.SIGTERM))
 	if err != nil {
 		logrus.Errorf("Sending SIGTERM to plugin failed with error: %v", err)
 	} else {
+
+		timeout := time.NewTimer(shutdownTimeout)
+		defer timeout.Stop()
+
 		select {
 		case <-ec:
 			logrus.Debug("Clean shutdown of plugin")
-		case <-time.After(time.Second * 10):
+		case <-timeout.C:
 			logrus.Debug("Force shutdown plugin")
 			if err := executor.Signal(pluginID, int(unix.SIGKILL)); err != nil {
 				logrus.Errorf("Sending SIGKILL to plugin failed with error: %v", err)
 			}
+
+			timeout.Reset(shutdownTimeout)
```

Comment:
Here the timer has already triggered and it's been drained.

Reply:
I don't think so. There is a select with two cases, one being this timer, so if the other case happens first, the timer is still active and ticking.

Reply:
`timer.Stop()` is called in a defer above.
This reset happens in the `<-timeout.C` branch.

Reply:
`defer` is executed later, and `<-timeout.C` branch might not be executed in case `<-ec` branch is first.

OK here is what I mean, let's take a look at the code flow:

```golang
		timeout := time.NewTimer(shutdownTimeout)
￼ 		defer timeout.Stop()
￼
￼ 		select {
￼ 		case <-ec:
￼ 			logrus.Debug("Clean shutdown of plugin")
￼ 		case <-time.After(time.Second * 10):
￼ 			logrus.Debug("Force shutdown plugin")
....
￼ 			}
￼
￼ 			timeout.Reset(shutdownTimeout)
```

Now, in case of clean shutdown (`case <-ec:` worked), the timer is not drained, and so `Reset()` works on a non-drained timer, which is racy as described in https://golang.org/pkg/time/#Timer.Reset.

Reply:
I am not seeing what you are seeing:

```go
		timeout := time.NewTimer(shutdownTimeout)
		defer timeout.Stop()

		select {
		case <-ec:
			logrus.Debug("Clean shutdown of plugin")
		case <-timeout.C:
			logrus.Debug("Force shutdown plugin")
			if err := executor.Signal(pluginID, int(unix.SIGKILL)); err != nil {
				logrus.Errorf("Sending SIGKILL to plugin failed with error: %v", err)
			}

			timeout.Reset(shutdownTimeout)

			select {
			case <-ec:
				logrus.Debug("SIGKILL plugin shutdown")
			case <-timeout.C:
				logrus.WithField("plugin", p.Name).Warn("Force shutdown plugin FAILED")
			}
		}
```

I do not see where reset is used on a non-drained timer.
This is the end of the function.

Reply:
of course! the second `select` is inside the first select's `case <-timeout.C:` so we're good here.